middleman_protocol/tests/tests.py

```diff
@@ -0,0 +1,453 @@
+            bytes(bytearray([100])) +
+            raw_message[payload_type_position + 1:]
+        )
+
+        # Replace message signature
+        new_signature = ecdsa_sign(CONCENT_PRIVATE_KEY, raw_message[FRAME_SIGNATURE_BYTES_LENGTH:])
+        raw_message_with_new_signature = new_signature + raw_message[FRAME_SIGNATURE_BYTES_LENGTH:]
+
+        with pytest.raises(MiddlemanProtocolError) as expected_exception:
+            AbstractMiddlemanMessage.deserialize(raw_message_with_new_signature, CONCENT_PUBLIC_KEY)
+
+        assertpy.assert_that(
+            str(expected_exception.value)
+        ).is_equal_to(
+            'Payload type 100 is not valid or not registered.'
+        )
+
+
+class TestHandleMessageReceiveHelper:
+
+    request_id = 99
+
+    @classmethod
+    def _replace_payload_length_and_payload_to_length_one(cls, raw_message):
+        return (
+            raw_message[:FRAME_PAYLOAD_LENGTH_STARTING_BYTE] +
+            b'\x00\x00\x00\x01' +
+            bytes(bytearray([0]))
+        )
+
+    @pytest.mark.parametrize(('use_separator', 'use_encoding'), [
+        (False, False),
+        (False, True),
+        (True, False),
+        (True, True),
+    ])
+    @pytest.mark.parametrize(('middleman_message_type', 'payload'), [
+        (GolemMessageMiddlemanMessage,            Ping()),
+        (ErrorMiddlemanMessage,                   (111, 'error_message')),
+        (AuthenticationChallengeMiddlemanMessage, b'random_bytes'),
+        (AuthenticationResponseMiddlemanMessage,  b'TODO'),
+    ])
+    def test_that_receiving_any_message_should_be_handled_correctly(
+        self,
+        middleman_message_type,
+        payload,
+        use_separator,
+        use_encoding,
+    ):
+        middleman_message = middleman_message_type(payload, self.request_id)
+        raw_message = middleman_message.serialize(
+            private_key=CONCENT_PRIVATE_KEY,
+            add_frame_separator=use_separator,
+        )
+
+        if use_encoding:
+            raw_message_ready_to_send = AbstractMiddlemanMessage.escape_encode_raw_message(raw_message)
+        else:
+            raw_message_ready_to_send = raw_message
+
+        with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as server_socket:
+            server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as client_socket:
+                client_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+                server_socket.bind(('127.0.0.1', 8001))
+                server_socket.listen(1)
+
+                client_socket.connect(('127.0.0.1', 8001))
+
+                (connection, _address) = server_socket.accept()
+
+                client_socket.send(raw_message_ready_to_send)
+                raw_message_received = next(handle_message_receive(connection, encoding_expected=use_encoding))
+
+        assertpy.assert_that(
+            str(raw_message)[2:-1]
```

Comment:
why `[2:-1]`?

Reply:
It seems that `assertpy` can't do `.startswith()` with bytes properly, so they have to be cast to strings first. These values cut of starting `b"` and finishing `"`. But if you know better way to do this with `assertpy` or `pytest` then off course it is worth changing.

Reply:
Maybe create a helper function to make it more obvious, e.g. `assertpy_starts_with()`?

Reply:
:+1: done

Reply:
@rwrzesien, why not just:
```python
 assertpy.assert_that(
            str(raw_message)
        ).starts_with(
            str(raw_message_received)
        )
```

Reply:
Beacuse `b"\x00\x00"` do not starts with `b"\x00"`.